### PR TITLE
6 add unit test for crud user and room

### LINKED ISSUE #6: Add unit test for crud user and room

### PR DESCRIPTION
**Pull request by Brawdunoir**

### CODE REVIEW

File: tests/room_test.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+var roomCreateRequest = utils.Request{Method: http.MethodPost, URL: roomURL, Body: `{"name":"test"}`}
+
+// TestCreateRoom tests the CreateRoom function.
+func TestCreateRoom(t *testing.T) {
+	method := http.MethodPost
+
+	test := utils.TestCreate{
+		SubTests: []utils.SubTest{
+			{Name: "Success", Request: utils.Request{Method: method, URL: roomURL, Body: `{"name":"test"}`}, ResponseCode: http.StatusCreated, ResponseBodyRegex: `{"id":"\d+"}`},
+			{Name: "Empty body", Request: utils.Request{Method: method, URL: roomURL, Body: ``}, ResponseCode: http.StatusBadRequest, ResponseBodyRegex: `{"error":".+"}`},
```
Comment (vazanoir) [inline tests/room_test.go:32]:
```suggestion
			{Name: "Empty body", Request: utils.Request{Method: method, URL: roomURL, Body: ``}, ResponseCode: http.StatusBadRequest, ResponseBodyRegex: `{"error":".+"}`},
			{Name: "Empty json", Request: utils.Request{Method: method, URL: roomURL, Body: `{}`}, ResponseCode: http.StatusBadRequest, ResponseBodyRegex: `{"error":".+"}`},
```

Comment (vazanoir) [reply tests/room_test.go:32]: Add it for every test like this

File: tests/room_test.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+			{Name: "Bad name key", Request: utils.Request{Method: method, URL: roomURL, Body: `{"wrongkey":"test"}`}, ResponseCode: http.StatusBadRequest, ResponseBodyRegex: `{"error":".+"}`},
+			{Name: "Bad name value", Request: utils.Request{Method: method, URL: roomURL, Body: `{"name":""}`}, ResponseCode: http.StatusBadRequest, ResponseBodyRegex: `{"error":".+"}`},
```
Comment (vazanoir) [inline tests/room_test.go]: Add tests for int/object/(nil/null) in values

**Review by vazanoir — CHANGES_REQUESTED** [review]
Moke rooms and users for every crud tests

File: tests/room_test.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+			{Name: "Empty name value", Request: utils.Request{Method: method, URL: roomURL, Body: `{"name":""}`}, ResponseCode: http.StatusBadRequest, ResponseBodyRegex: `{"error":".+"}`},
+			{Name: "Nil name value", Request: utils.Request{Method: method, URL: roomURL, Body: `{"name":nil}`}, ResponseCode: http.StatusBadRequest, ResponseBodyRegex: `{"error":".+"}`},
+			{Name: "Integer name value", Request: utils.Request{Method: method, URL: roomURL, Body: `{"name":1}`}, ResponseCode: http.StatusBadRequest, ResponseBodyRegex: `{"error":".+"}`},
+			{Name: "Object name value", Request: utils.Request{Method: method, URL: roomURL, Body: `{"name":{"somekey":"somevalue"}}`}, ResponseCode: http.StatusBadRequest, ResponseBodyRegex: `{"error":".+"}`},
+		},
+	}
+	test.Run(t)
+}
+
+// TestGetRoom test the GetRoom function.
```
Comment (WhiteDiamondz) [inline tests/room_test.go]:
```suggestion
// TestGetRoom tests the GetRoom function.
```

File: tests/room_test.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+		CreateResponse: CreateResponseRoom{},
+		SubTests: []utils.SubTest{
+			{Name: "Success", Request: utils.Request{Method: method, URL: roomURL}, ResponseCode: http.StatusOK, ResponseBodyRegex: `{"room":{"name":"test"}}`},
+			{Name: "Not found", Request: utils.Request{Method: method, URL: roomURL + "0"}, ResponseCode: http.StatusNotFound, ResponseBodyRegex: `{"error":"Room not found"}`},
+		},
+	}
+	test.Run(t)
+}
+
+// TestUpdateRoom test the UpdateRoom function.
```
Comment (WhiteDiamondz) [inline tests/room_test.go]:
```suggestion
// TestUpdateRoom tests the UpdateRoom function.
```

File: tests/room_test.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+			{Name: "Nil name value", Request: utils.Request{Method: method, URL: roomURL, Body: `{"name":nil}`}, ResponseCode: http.StatusBadRequest, ResponseBodyRegex: `{"error":".+"}`},
+			{Name: "Integer name value", Request: utils.Request{Method: method, URL: roomURL, Body: `{"name":1}`}, ResponseCode: http.StatusBadRequest, ResponseBodyRegex: `{"error":".+"}`},
+			{Name: "Object name value", Request: utils.Request{Method: method, URL: roomURL, Body: `{"name":{"somekey":"somevalue"}}`}, ResponseCode: http.StatusBadRequest, ResponseBodyRegex: `{"error":".+"}`},
+			{Name: "Not found", Request: utils.Request{Method: method, URL: roomURL + "0", Body: `{"name":"test2"}`}, ResponseCode: http.StatusNotFound, ResponseBodyRegex: `{"error":"Room not found"}`},
+		},
+	}
+	test.Run(t)
+}
+
+// TestDeleteRoom test the DeleteRoom function.
```
Comment (WhiteDiamondz) [inline tests/room_test.go]:
```suggestion
// TestDeleteRoom tests the DeleteRoom function.
```

File: tests/user_test.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+			{Name: "Empty username value", Request: utils.Request{Method: method, URL: userURL, Body: `{"username":""}`}, ResponseCode: http.StatusBadRequest, ResponseBodyRegex: `{"error":".+"}`},
+			{Name: "Nil username value", Request: utils.Request{Method: method, URL: userURL, Body: `{"username":nil}`}, ResponseCode: http.StatusBadRequest, ResponseBodyRegex: `{"error":".+"}`},
+			{Name: "Integer username value", Request: utils.Request{Method: method, URL: userURL, Body: `{"username":1}`}, ResponseCode: http.StatusBadRequest, ResponseBodyRegex: `{"error":".+"}`},
+			{Name: "Object username value", Request: utils.Request{Method: method, URL: userURL, Body: `{"username":{"somekey":"somevalue"}}`}, ResponseCode: http.StatusBadRequest, ResponseBodyRegex: `{"error":".+"}`},
+		},
+	}
+	test.Run(t)
+}
+
+// TestGetUser test the GetUser function.
```
Comment (WhiteDiamondz) [inline tests/user_test.go]:
```suggestion
// TestGetUser tests the GetUser function.
```

File: tests/user_test.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+		CreateResponse: CreateResponseUser{},
+		SubTests: []utils.SubTest{
+			{Name: "Success", Request: utils.Request{Method: method, URL: userURL}, ResponseCode: http.StatusOK, ResponseBodyRegex: `{"user":{"username":"test"}}`},
+			{Name: "Not found", Request: utils.Request{Method: method, URL: userURL + "0"}, ResponseCode: http.StatusNotFound, ResponseBodyRegex: `{"error":"User not found"}`},
+		},
+	}
+	test.Run(t)
+}
+
+// TestUpdateUser test the UpdateUser function.
```
Comment (WhiteDiamondz) [inline tests/user_test.go]:
```suggestion
// TestUpdateUser tests the UpdateUser function.
```

File: tests/user_test.go [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+			{Name: "Nil username value", Request: utils.Request{Method: method, URL: userURL, Body: `{"username":nil}`}, ResponseCode: http.StatusBadRequest, ResponseBodyRegex: `{"error":".+"}`},
+			{Name: "Integer username value", Request: utils.Request{Method: method, URL: userURL, Body: `{"username":1}`}, ResponseCode: http.StatusBadRequest, ResponseBodyRegex: `{"error":".+"}`},
+			{Name: "Object username value", Request: utils.Request{Method: method, URL: userURL, Body: `{"username":{"somekey":"somevalue"}}`}, ResponseCode: http.StatusBadRequest, ResponseBodyRegex: `{"error":".+"}`},
+			{Name: "Not found", Request: utils.Request{Method: method, URL: userURL + "0", Body: `{"username":"test2"}`}, ResponseCode: http.StatusNotFound, ResponseBodyRegex: `{"error":"User not found"}`},
+		},
+	}
+	test.Run(t)
+}
+
+// TestDeleteUser test the DeleteUser function.
```
Comment (WhiteDiamondz) [inline tests/user_test.go]:
```suggestion
// TestDeleteUser tests the DeleteUser function.
```

File: utils/tests/tests.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+type ITest interface {
+	Run(t *testing.T)
+}
+
+// ICreateResponse allows to map a Create request and retrieve the key for further tests.
+type ICreateResponse interface {
+	KeyCreated([]byte) (string, error)
+}
+
+// SubTest is an atomic test that include a request and its intended response.
```
Comment (WhiteDiamondz) [inline utils/tests/tests.go]:
```suggestion
// SubTest is an atomic test that includes a request and its intended response.
```

File: utils/tests/tests.go [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+type SubTest struct {
+	Name              string
+	Request           Request
+	ResponseCode      int
+	ResponseBodyRegex interface{}
+}
+
+// Request simple request to be send to the router.
```
Comment (WhiteDiamondz) [inline utils/tests/tests.go]:
```suggestion
// Request a simple request to be sent to the router.
```

File: utils/tests/tests_create.go [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+	"github.com/go-playground/assert/v2"
+)
+
+// TestCreate struct to factorize code inside a single feature test.
+type TestCreate struct {
+	SubTests []SubTest
+}
+
+// Runs a series of tests for a Create type endpoint.
```
Comment (WhiteDiamondz) [inline utils/tests/tests_create.go]:
```suggestion
// Runs a series of tests for the Create type endpoint.
```

File: utils/tests/tests_rud.go [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+)
+
+// TestRUD (Read, Update, Delete) struct to factorize code inside a single feature test.
+type TestRUD struct {
+	CreateRequest  Request
+	CreateResponse ICreateResponse
+	SubTests       []SubTest
+}
+
+// Runs a series of tests for a Get type endpoint.
```
Comment (WhiteDiamondz) [inline utils/tests/tests_rud.go]:
```suggestion
// Runs a series of tests for the Get type endpoint.
```

**Review by WhiteDiamondz — CHANGES_REQUESTED** [review]
Small fixes in comments, otherwise good to merge once corrected !

**Review by WhiteDiamondz — APPROVED** [review]
Good to go !

**Review by vazanoir — APPROVED** [review]
